### ISSUE TITLE
Flickering on scaled ultrawide monitor on M3 Pro Mac

### ISSUE THREAD

**Issue by matthieusieben** [microsoft/vscode#219660]
Type: <b>Bug</b>

Random parts of the VS Code window flicker when typing / scrolling.

https://github.com/microsoft/vscode/assets/813661/f208c708-0c2f-429b-870c-b9a738dfa142

https://github.com/microsoft/vscode/assets/813661/4cb1e107-5b78-4747-8cd3-0d292bd40c73

https://github.com/microsoft/vscode/assets/813661/bd6d8993-6f25-4f2c-8bfc-bbd77f677d2a

Steps to Reproduce:

1. External screen with native resolution of 5120x2160 (Dell U4025QW)
2. Setup scaled resolution of 3840x1620. This is OS scaling, the monitor still reports a native resolution of 5120x2160.
3. Open VSCode with all extensions disabled, and default settings (json settings emptied).
4. Open several files side by side, optionally using "diff mode". The more visible editors, the more the error is likely to appear.

Note:

- This is not linked to dark colors as described in other issues. (Issue occurs when OS is in light mode, with a light theme)
- Scrolling while VS Code is *not* focussed does not trigger the issue
- Happens both in full screen mode and non-fullscreen mode

Similar issues:

- https://github.com/microsoft/vscode/issues/193908
- https://github.com/microsoft/vscode/issues/165681

Workaround:

- Disable HiDPI scaling in OS and use VS Code's window zoom. This affects all apps, making it not suitable.
- Set `"disable-hardware-acceleration": true,` in  `Preferences: Configure Runtime Arguments`.

VS Code version: Code 1.90.2 (5437499feb04f7a586f677b155b039bc2b3669eb, 2024-06-18T22:54:35.075Z)
OS version: Darwin arm64 23.5.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M3 Pro (12 x 2400)|
|GPU Status|2d_canvas: unavailable_software<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: disabled_software<br>multiple_raster_threads: enabled_on<br>opengl: disabled_off<br>rasterization: disabled_software<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: disabled_software<br>video_encode: disabled_software<br>webgl: unavailable_software<br>webgl2: unavailable_software<br>webgpu: unavailable_software|
|Load (avg)|2, 2, 2|
|Memory (System)|36.00GB (0.30GB free)|
|Process Argv|--crash-reporter-id b9f963fa-db84-4aac-9a3d-85fa771e9805|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (28)</summary>

Extension|Author (truncated)|Version
---|---|---
lit-html|bie|1.11.1
vscode-tailwindcss|bra|0.12.3
npm-intellisense|chr|1.4.5
postcss|css|1.0.9
vscode-eslint|dba|3.0.10
eslint-disable-snippets|drK|1.4.1
gitlens|eam|15.1.0
EditorConfig|Edi|0.16.4
prettier-vscode|esb|10.4.0
vscode-firefox-debug|fir|2.9.10
vscode-jest-runner|fir|0.4.73
copilot|Git|1.208.0
copilot-chat|Git|0.16.1
copy-github-url|mat|0.14.0
import-sorter|mik|3.3.1
dotenv|mik|1.0.1
ecdc|mit|1.8.0
hexeditor|ms-|1.10.0
sublime-keybindings|ms-|4.1.10
vscode-typescript-next|ms-|5.6.20240701
one-dark-theme|msk|1.14.2
material-icon-theme|PKi|5.4.2
rewrap|stk|17.8.0
code-spell-checker|str|3.0.1
sort-lines|Tyr|1.11.0
highlight-bad-chars|wen|0.0.6
change-case|wma|1.0.0
markdown-all-in-one|yzh|3.6.2

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscoreces:30445986
vscod805cf:30301675
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonnoceb:30805159
asynctok:30898717
pythonregdiag2:30936856
pythonmypyd1:30879173
2e7ec940:31000449
pythontbext0:30879054
accentitlementsc:30995553
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
jchc7451:31067544
chatpanelc:31048052
dsvsc021:30996838
g316j359:31013175
pythoncenvpt:31062603
a69g1124:31058053
dvdeprecation:31068756
dwnewjupytercf:31046870
newcmakeconfigv2:31071590
impr_priority:31082723

```

</details>

**Comment by matthieusieben** [comment]
Tried on `https://vscode.dev` on the same project, but the flickering did not happen there:

https://github.com/microsoft/vscode/assets/813661/1c7006f6-626b-4fff-b915-53292c372836

**Comment by matthieusieben** [comment]
Also, opening more panels seem to make the issue worse but it also happens when only one file is open:


https://github.com/microsoft/vscode/assets/813661/9df8d898-c142-47f1-bba3-c2aaf3d60ff5

**Comment by matthieusieben** [comment]
The issue appears to be linked to the resolution of the window and, in my setup, happens significantly more when the window goes above approx 6200x3192. Note that the window resolution is significantly higher than my screen resolution (5120x2160) probably because of apple's HiDPI feature.


<img width="3088" alt="No flickering at this resolution" src="https://github.com/microsoft/vscode/assets/813661/9bcb1af6-8038-430e-95e1-1c9904ebfa20">

**Comment by zekedroid** [comment]
Having the same issue under the same conditions: M3 Pro Mac, full screen editor windows. It helps to manually zoom the VSCode instance but it's non-deterministic.

**Comment by deepak1556** [comment]
Can you confirm if the issue persists with latest stable, if it does please share the following outputs

1) `chrome://gpu` from your chrome browser
2) `Developer: Show GPU info` from VSCode

**Comment by matthieusieben** [comment]
The flickering started happening again (on v1.96.2).

It seems that what triggers the issue is when the window is first opened on my laptop's screen, and is then moved to a to the secondary monitor (and put in full screen mode).

Here is why I think that is the case:

If I close my laptop, forcing the initial window to open on my external monitor, the flickering won't happen. Even if I move the window back and forth between my monitors.

If I use the "Duplicate Workspace in New Window" command, while in full screen on my external monitor (with a flickering instance), the new instance (which gets opened on the secondary monitor) does not flicker.

If I then open a "New Window" (right click on dock) while both screens are active, the initial windows does not flicker (if opened from external monitor) but the new one does (when moved to fullscreen on external monitor).

Just mentioning in case it matters: When I launch VScode with the computer lid closed (and the flickering does not happen), it looks like my external monitor switches to another color profile (the colors on the external monitor are visibly "bluer" when I close my laptop). The system preferences do not show a different color profile being used.

Here are the requested files:
- [Google Chrome](https://github.com/user-attachments/files/18331830/about-gpu-2025-01-07T11-28-34-642Z.txt)
- [VSCode](https://github.com/user-attachments/files/18331833/about-gpu-2025-01-07T11-28-14-084Z.txt) flickering
- [VScode](https://github.com/user-attachments/files/18332023/about-gpu-2025-01-07T11-44-47-838Z.txt) no flickering (app started with laptop closed, on external monitor, gpu info taken with lid closed)
- [VScode](https://github.com/user-attachments/files/18332241/about-gpu-2025-01-07T12-02-19-361Z.txt) no flickering (app started with laptop closed, on external monitor, gpu info taken with lid open again)
- [VScode](https://github.com/user-attachments/files/18332140/about-gpu-2025-01-07T11-53-56-036Z.txt) flickering (launched from flickering "new window", with initial window, launched while lid was closed, not flickering)

**Comment by matthieusieben** [comment]
As it stands there are two workarounds:
- Disable GPU acceleration
- Open VSCode (will open on laptop monitor by default). Move initial (buggy) window on external screen. Open "New Window" (CMD+shift+N). Close initial window. Use the window created from external screen.

**Comment by matthieusieben** [comment]
I think I know why it started happening again recently. It seems that VScode remembers the "last used monitor" on launch, but it will only remember the last monitor if the last window was closed in window mode.

Because I used my laptop without external screen, the "initial monitor" was reset to being my laptop monitor, which caused the error to happen again.

Closing the vscode window (while not in full screen) from my external monitor allowed me to start opening non buggy initial windows from the external monitor.

tl;dr: the issue was probably never really fixed. It [appeared](https://github.com/microsoft/vscode/issues/219660#issuecomment-2553431958) fixed for me only because the window were launching on the right monitor.

**Comment by matthieusieben** [comment]
I tried again with versions 1.96.1, 1.96.0 & 1.95.3 and the behavior is indeed the same as described in my [previous comment](https://github.com/microsoft/vscode/issues/219660#issuecomment-2575130834).
